### PR TITLE
Fix #2046: Bone rotating bugs

### LINKED ISSUE #2046: Bone rotating bugs
**1. Bone's rotation doesn't work with CJ skin.**
*Just use example code from [wiki page](https://wiki.multitheftauto.com/wiki/SetElementBoneRotation) and change skin ids using default admin resource.*

**2. Weapon position and rotation don't take into account bone's position and rotation.**
*This bug cause wrong shot position and it's direction. As the result it affects on functions such as [getPedTargetStart](https://wiki.multitheftauto.com/wiki/GetPedTargetStart) and [getPedTargetEnd](https://wiki.multitheftauto.com/wiki/GetPedTargetEnd). This probably requires additional functions.*

*Screenshot*

**image**

*Code*
```Lua
addEventHandler ( "onClientPedsProcessed", root, function ( )
	if localPlayer:getTask ( 'secondary', 0 ) ~= 'TASK_SIMPLE_USE_GUN' then return end -- may be deleted
	local bone = 2 -- Pelvis
	local rot = 30
	local yaw, pitch, roll = getElementBoneRotation ( localPlayer, bone )
	setElementBoneRotation ( localPlayer, bone, yaw, pitch - rot, roll + rot )
	updateElementRpHAnim ( localPlayer )
end )

addEventHandler("onClientPreRender", root,  -- Adds the Handler
    function ()
        if getPedTargetStart(localPlayer) then --Checks if there is a Point to start From.

        local x, y, z = getPedTargetStart(localPlayer) -- Gets the Point to start From
        local sx, sy, sz = getPedTargetEnd(localPlayer) -- Gets the Point where the Target Ends

        dxDrawLine3D(x, y, z, sx, sy, sz) -- Draws the Line
    end
end
)
```
*Client: v1.5.8-release-20779
Server: v1.5.8-release-20817*

### PR DESCRIPTION
**Pull request by jey-banned**
fix #2046:
- Fixed CJ's shoulders handling
- Note added to wiki (aiming functions)

Big thanks to @ghost with explanations and coding.

### CODE REVIEW

File: Client/game_sa/CEntitySA.cpp [hunk 1]
```diff
@@ -678,6 +678,9 @@ bool CEntitySA::SetBoneRotation(eBone boneId, float yaw, float pitch, float roll
             RtQuat* boneOrientation = &frameData->m_pIFrame->orientation;
             RwV3d   angles = {yaw, roll, pitch};
             BoneNode_cSAInterface::EulerToQuat(&angles, boneOrientation);
+            CEntitySAInterface* theInterface = GetInterface();
+            if (theInterface)
```
Comment (se16n) [inline Client/game_sa/CEntitySA.cpp:682]: I think blocks without braces should be generally avoided.

File: Client/game_sa/CEntitySA.cpp [hunk 2]
```diff
@@ -678,6 +678,9 @@ bool CEntitySA::SetBoneRotation(eBone boneId, float yaw, float pitch, float roll
...
+                theInterface->bDontUpdateHierarchy = false;
```
Comment (se16n) [inline Client/game_sa/CEntitySA.cpp]: I think blocks without braces should be generally avoided.

File: Client/mods/deathmatch/logic/luadefs/CLuaPedDefs.cpp [hunk 3]
```diff
@@ -1040,6 +1040,11 @@ bool CLuaPedDefs::UpdateElementRpHAnim(lua_State* const luaVM, CClientEntity* en
     if (theEntity)
     {
         theEntity->UpdateRpHAnim();
+        if (theEntity->GetModelIndex() == 0) { // CJ skin
+            RpClump* clump = theEntity->GetRpClump();
+            if (clump)
+                ((void(__cdecl*)(RpClump*))0x5DF560)(clump); // CPed::ShoulderBoneRotation
```
Comment (se16n) [inline Client/mods/deathmatch/logic/luadefs/CLuaPedDefs.cpp]: I think blocks without braces should be generally avoided.

Comment (jey-banned) [reply Client/mods/deathmatch/logic/luadefs/CLuaPedDefs.cpp]: you're right, thank you

File: Client/game_sa/CEntitySA.cpp [hunk 4]
```diff
@@ -678,6 +678,10 @@ bool CEntitySA::SetBoneRotation(eBone boneId, float yaw, float pitch, float roll
             RtQuat* boneOrientation = &frameData->m_pIFrame->orientation;
             RwV3d   angles = {yaw, roll, pitch};
             BoneNode_cSAInterface::EulerToQuat(&angles, boneOrientation);
+            CEntitySAInterface* theInterface = GetInterface();
+            if (theInterface) {
+                theInterface->bDontUpdateHierarchy = false;
+            }
```
Comment (Pirulax) [inline Client/game_sa/CEntitySA.cpp]:
```cpp
if (GetInterface())
    GetInterface()->bDontUpdateHierarchy  = false;
```
You dont really need the braces. If you do chose to use them, you must put them in a new line.

**Review by Pirulax — APPROVED** [review]

File: Client/game_sa/CEntitySA.cpp (hunk 4 above)
Comment (jey-banned) [reply Client/game_sa/CEntitySA.cpp]: Thank you for noticing. I'll be carefull next time

**Comment by Pirulax** [comment]
Have you tested it, does it work as expected? If so I'll merge it.

**Comment by jey-banned** [comment]
Yep. onClientPedsProcessed event works with CJ's skin now as expected.

**Review by patrikjuvonen — APPROVED** [review]

**Comment by patrikjuvonen** [comment]
Thank you for the pull request and congratulations on your first-time contribution! 🎉

**Comment by Pirulax** [comment]
Amazing. Thanks a lot!

**Comment by FuccDucc** [comment]
> 
> 
> Thank you for the pull request and congratulations on your first-time contribution! 🎉

He didn't contribute.

As he didn't write a single line of code by himself. All the code from this patch was written by someone else in 
_Originally posted by @ghost in https://github.com/multitheftauto/mtasa-blue/issues/2046#issuecomment-771840094_

Jevico looked at the image and typed over the text.

**Pull request merged**